File: JavaScript/SubarraySumEqualsK/SubarraySumEqualsK.js
/**
 * (299ms)
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function(nums, k) {
    var count = 0, nowSum = 0;
    var prevSum = new Map();
    prevSum.set(0, 1);
    for(var num of nums) {
    	nowSum += num;
    	if(prevSum.has(nowSum - k)) {
    		count += prevSum.get(nowSum - k);
    	}
    	if(prevSum.has(nowSum)) {
    		prevSum.set(nowSum, prevSum.get(nowSum) + 1);
    	} else {
    		prevSum.set(nowSum, 1);
    	}
    } 
    return count;
};

File: Python/SubarraySumEqualsK/SubarraySumEqualsK.py
class Solution(object):
	# (62ms)
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        count, nowSum, prevSum = 0, 0, {0 : 1}
        for num in nums:
        	nowSum += num
        	if nowSum - k in prevSum:
        		count += prevSum[nowSum - k]
        	prevSum[nowSum] = prevSum[nowSum] + 1 if nowSum in prevSum else 1
        return count

print Solution().subarraySum([1, 1, 1], 2)